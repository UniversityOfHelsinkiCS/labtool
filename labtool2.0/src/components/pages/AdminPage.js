import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Button, Loader } from 'semantic-ui-react'
import { Redirect } from 'react-router'
import { getAllCI } from '../../services/courseInstance'
import { resetLoading } from '../../reducers/loadingReducer'
import { getAllUsers, updateOtherUser } from '../../services/user'
import { clearNotifications } from '../../reducers/notificationReducer'
import { usePersistedState } from '../../hooks/persistedState'
import { AdminCourseList } from './AdminPage/AdminCourseList'
import { AdminUserList } from './AdminPage/AdminUserList'
import DocumentTitle from '../DocumentTitle'
import Error from '../Error'

export const AdminPage = props => {
  const state = usePersistedState('AdminPage', {
    coursesList: false,
    usersList: false
  })

  useEffect(() => {
    props.clearNotifications()
  }, [])

  useEffect(() => {
    if (state.coursesList) {
      props.resetLoading()
      props.getAllCI()
    }
  }, [state.coursesList])

  useEffect(() => {
    if (state.usersList) {
      props.resetLoading()
      props.getAllUsers()
    }
  }, [state.usersList])

  const documentTitle = <DocumentTitle title={'Admin'} />

  if (!props.user.user.sysop) {
    return <Redirect to={`/labtool`} />
  } else if (props.errors && props.errors.length > 0) {
    return <Error errors={props.errors.map(error => `${error.response.data} (${error.response.status} ${error.response.statusText})`)} />
  } else if (props.loading.loading) {
    return (
      <>
        <Loader active />
        {documentTitle}
      </>
    )
  }

  return (
    <>
      {documentTitle}
      <div className="AdminPage">
        <h2>Welcome to the admin interface.</h2>
        <Button compact className={`ui button`} onClick={() => (state.coursesList = !state.coursesList)}>
          {state.coursesList ? 'Hide courses' : 'Manage courses'}
        </Button>
        <br />
        <br />
        {state.coursesList && <AdminCourseList courses={props.courseInstance} />}
        <br />
        <hr />
        <br />
        <Button compact className={`ui button`} onClick={() => (state.usersList = !state.usersList)}>
          {state.usersList ? 'Hide users' : 'Manage users'}
        </Button>
        <br />
        {state.usersList && <AdminUserList me={props.user} users={props.users} updateOtherUser={props.updateOtherUser} />}
      </div>
    </>
  )
}

const mapStateToProps = state => {
  return {
    user: state.user,
    courseInstance: state.courseInstance,
    users: state.users,
    loading: state.loading,
    errors: Object.values(state.loading.errors)
  }
}

const mapDispatchToProps = {
  getAllCI,
  resetLoading,
  getAllUsers,
  updateOtherUser,
  clearNotifications
}

AdminPage.propTypes = {
  user: PropTypes.object.isRequired,
  courseInstance: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  users: PropTypes.array.isRequired,
  loading: PropTypes.object.isRequired,

  getAllCI: PropTypes.func.isRequired,
  resetLoading: PropTypes.func.isRequired,
  getAllUsers: PropTypes.func.isRequired,
  updateOtherUser: PropTypes.func.isRequired,
  clearNotifications: PropTypes.func.isRequired,

  errors: PropTypes.array
}

export default connect(mapStateToProps, mapDispatchToProps)(AdminPage)
