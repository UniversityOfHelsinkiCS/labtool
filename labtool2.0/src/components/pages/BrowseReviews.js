import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router'
import { Button, Card, Accordion, Icon, Form, Comment, Input, Popup, Loader, Label, Header } from 'semantic-ui-react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { createOneComment } from '../../services/comment'
import { getOneCI, coursePageInformation } from '../../services/courseInstance'
import { gradeCodeReview } from '../../services/codeReview'
import { updateStudentProjectInfo } from '../../services/studentinstances'
import ReactMarkdown from 'react-markdown'
import { sendEmail } from '../../services/email'
import { resetLoading } from '../../reducers/loadingReducer'
import { createCourseIdWithYearAndTerm } from '../../util/format'
import { getCoursesByStudentId } from '../../services/studentinstances'
import useLegacyState from '../../hooks/legacyState'

import BackButton from '../BackButton'
import LabtoolComment from '../LabtoolComment'
import LabtoolAddComment from '../LabtoolAddComment'
import { formatCourseName } from '../../util/format'
import RepoLink from '../RepoLink'

/**
 * Maps all comments from a single instance from coursePage reducer
 */
export const BrowseReviews = props => {
  const state = useLegacyState({
    openWeeks: {},
    initialLoading: props.initialLoading !== undefined ? props.initialLoading : true
  })
  let jumpTimer = null

  useEffect(() => {
    // run on component mount
    props.resetLoading()
    props.getOneCI(props.courseId)
    props.coursePageInformation(props.courseId)
    props.getCoursesByStudentId(Number(props.studentInstance))
    if (!props.loading.loading && !state.openWeeks[props.selectedInstance.currentWeek - 1]) {
      state.openWeeks = { [props.selectedInstance.currentWeek - 1]: true }
    }

    if (props.location.state && props.location.state.jumpToReview !== undefined) {
      // jumpToReview from <Link state={...}>: scrolls to review
      const tryJumpLoop = () => {
        if (!tryJumpToReview()) {
          // not loaded yet? try again later
          jumpTimer = setTimeout(tryJumpLoop, 200)
        }
      }
      jumpTimer = setTimeout(tryJumpLoop, 200)
    }

    return () => {
      clearTimeout(jumpTimer)
    }
  }, [])

  useEffect(() => {
    if (!props.loading.loading && state.initialLoading) {
      state.initialLoading = false
    }
  }, [props.loading.loading, state.initialLoading])

  useEffect(() => {
    if (props.location.state && props.location.state.openAllWeeks) {
      handleClickShowAllCurrent()
    }
  }, [props.courseData.data])

  const tryJumpToReview = () => {
    // Try scroll to review
    const element = document.getElementById(`review${props.location.state.jumpToReview}`)

    if (element) {
      window.scrollTo(0, element.offsetTop)
      return true
    }
    return false
  }

  const handleClick = (e, titleProps) => {
    const { index } = titleProps
    const { openWeeks } = state
    const numberOfOpen = Object.values(openWeeks).filter(x => x).length

    if (numberOfOpen === 0) {
      // if no reviews open, open the clicked tab
      state.openWeeks = { [index]: true }
    } else if (numberOfOpen === 1) {
      // if one open, open the clicked tab and close everything else,
      // or close everything if we clicked the open one
      state.openWeeks = { [index]: !openWeeks[index] }
    } else {
      // if multiple open, simply toggle open/close
      state.openWeeks = { ...openWeeks, [index]: !openWeeks[index] }
    }
  }

  const getMaximumIndexForStudent = student => {
    // how many reviews will there be?
    return props.selectedInstance.weekAmount + student.codeReviews.length + (props.selectedInstance.finalReview ? 2 : 1)
  }

  const hasAllReviewsOpen = student => {
    return Object.values(state.openWeeks).filter(x => x).length === getMaximumIndexForStudent(student)
  }

  const handleClickShowAll = student => () => {
    const openWeeks = {}
    const maximumIndex = getMaximumIndexForStudent(student)
    for (let i = 0; i < maximumIndex; i++) {
      openWeeks[i] = true
    }
    state.openWeeks = openWeeks
  }

  const handleClickShowAllCurrent = () => {
    if (props.courseData.data && props.studentInstance) {
      const student = props.courseData.data.find(student => student.id === Number(props.studentInstance))
      if (student) {
        handleClickShowAll(student)()
      }
    }
  }

  const handleClickHideAll = () => {
    state.openWeeks = {}
  }

  const handleSubmit = async e => {
    e.preventDefault()
    const content = {
      hidden: e.target.hidden.checked,
      comment: e.target.content.value,
      week: parseInt(e.target.name, 10)
    }
    document.getElementById(e.target.name).reset()
    try {
      await props.createOneComment(content)
    } catch (error) {
      console.error(error)
    }
  }

  const handleMarkAsDropped = async dropped => {
    await props.updateStudentProjectInfo({
      ohid: props.selectedInstance.ohid,
      userId: props.courseData.data.find(data => data.id === Number(props.studentInstance)).userId,
      dropped
    })
  }

  const sortCommentsByDate = comments => {
    return comments.sort((a, b) => {
      return new Date(a.createdAt) - new Date(b.createdAt)
    })
  }

  const gradeCodeReview = (reviewNumber, studentInstanceId) => async e => {
    e.preventDefault()
    const data = {
      reviewNumber,
      studentInstanceId: Number(studentInstanceId),
      points: Number(e.target.points.value)
    }
    props.gradeCodeReview(data)
  }

  const sendCommentEmail = commentId => async () => {
    props.sendEmail({
      commentId,
      role: 'teacher'
    })
  }

  const sendWeekEmail = weekId => async () => {
    props.sendEmail({
      weekId,
      role: 'teacher'
    })
  }

  const isTeacher = () => {
    return props.courseData.role === 'teacher'
  }

  //get student's other participations in the same course
  const renderStudentPreviousParticipation = () => {
    const previousParticipations = props.studentInstanceToBeReviewed.filter(courseInstance => courseInstance.ohid.includes(props.courseId.substring(0, 8)) && courseInstance.ohid !== props.courseId)
    if (previousParticipations.length === 0) {
      return <p className="noOther">Has no other participation in this course</p>
    }
    return (
      <div className="hasOther">
        <p style={{ color: 'red' }}>Has taken this course in other periods</p>
        {previousParticipations.map(participation => (
          <Link key={participation.id} to={`/labtool/browsereviews/${participation.ohid}/${participation.courseInstances[0].id}`}>
            <Popup content="click to see the details" trigger={<p>{createCourseIdWithYearAndTerm(participation.ohid, participation.start)}</p>} />
          </Link>
        ))}
      </div>
    )
  }

  const renderStudentCard = student => (
    <Card key={student.id} fluid color="yellow" className="studentCard">
      <Card.Content>
        <Header as="h2">
          {student.User.firsts} {student.User.lastname} ({student.User.studentNumber})
          {student.dropped && (
            <Label style={{ marginTop: 5, float: 'right' }}>
              <Icon name="warning sign" color="red" />
              Has dropped course
            </Label>
          )}
          <Header.Subheader>
            <a href={`mailto:${student.User.email}`}>{student.User.email}</a>
          </Header.Subheader>
        </Header>
        <div style={{ display: 'inline-block' }}>
          {student.projectName}: <RepoLink url={student.github} />
          <br />
          {renderStudentPreviousParticipation()}
        </div>
        {
          <Button color="red" style={{ float: 'right' }} onClick={() => handleMarkAsDropped(!student.dropped)}>
            {student.dropped ? 'Mark as non-dropped' : 'Mark as dropped'}
          </Button>
        }
      </Card.Content>
    </Card>
  )

  const renderComment = (isFinalWeek, comment) => {
    /* This hack compares user's name to comment.from and hides the email notification button when they don't match. */
    const userIsCommandSender = comment.from.includes(props.user.user.firsts) && comment.from.includes(props.user.user.lastname)

    return <LabtoolComment key={comment.id} comment={comment} allowNotify={!isFinalWeek && userIsCommandSender} sendCommentEmail={sendCommentEmail(comment.id)} />
  }

  const renderWeek = (i, week, studentInstance, isFinalWeek) => {
    const { openWeeks } = state
    const reviewIndex = isFinalWeek ? props.selectedInstance.weekAmount + 1 : i + 1

    if (week) {
      return (
        <Accordion fluid styled id={`review${i}`}>
          <Accordion.Title active={openWeeks[i]} index={i} onClick={handleClick}>
            <Icon name="dropdown" /> {isFinalWeek ? 'Final Review' : `Week ${i + 1}`}, points {week.points}
          </Accordion.Title>
          <Accordion.Content active={openWeeks[i]}>
            <h3>Review</h3>
            <Link to={`/labtool/reviewstudent/${props.selectedInstance.ohid}/${studentInstance}/${reviewIndex}`}>
              <Popup trigger={<Button circular color="orange" size="tiny" icon={{ name: 'edit', color: 'black', size: 'large' }} />} content={isFinalWeek ? 'Edit final review' : 'Edit review'} />
            </Link>
            <Card fluid color="yellow">
              <Card.Content>
                <h4> Points {week.points} </h4>
                <h4> Feedback </h4>
                <ReactMarkdown>{week.feedback}</ReactMarkdown>{' '}
                {isTeacher() && week.instructorNotes ? (
                  <div>
                    <br />
                    <h4>Review notes for instructors </h4>
                    <ReactMarkdown>{week.instructorNotes}</ReactMarkdown>{' '}
                  </div>
                ) : (
                  <span />
                )}
              </Card.Content>
              {!isFinalWeek ? (
                <Card.Content style={{ paddingBottom: '5px' }}>
                  {week.notified ? (
                    <Label>
                      Notified <Icon name="check" color="green" />
                    </Label>
                  ) : (
                    <Button type="button" onClick={sendWeekEmail(week.id)} size="small">
                      Send email notification
                    </Button>
                  )}
                </Card.Content>
              ) : (
                <span />
              )}
            </Card>
            <h4> Comments </h4>
            <Comment.Group>{week ? sortCommentsByDate(week.comments).map(c => renderComment(isFinalWeek, c)) : <h4> No comments </h4>}</Comment.Group>
            <LabtoolAddComment weekId={week.id} commentFieldId={`${props.courseId}:${week.id}`} handleSubmit={handleSubmit} allowHidden={true} />
          </Accordion.Content>
        </Accordion>
      )
    } else {
      return (
        <Accordion fluid styled id={`review${i}`}>
          <Accordion.Title active={openWeeks[i]} index={i} onClick={handleClick}>
            <Icon name="dropdown" /> {isFinalWeek ? 'Final Review' : `Week ${i + 1}`}{' '}
          </Accordion.Title>
          <Accordion.Content active={openWeeks[i]}>
            <h3>Review</h3>
            <Link to={`/labtool/reviewstudent/${props.selectedInstance.ohid}/${studentInstance}/${reviewIndex}`}>
              <Popup trigger={<Button circular color="orange" size="tiny" icon={{ name: 'edit', color: 'black', size: 'large' }} />} content={isFinalWeek ? 'Give Final Review' : 'Review week'} />
            </Link>
            <h4> Not Graded </h4>
            <h4> No comments </h4>
          </Accordion.Content>
        </Accordion>
      )
    }
  }

  // render link of the repo which is assigned to the student to review
  const renderAssignedCodeReviewLink = cr => {
    if (cr.toReview) {
      return (
        <p>
          <strong>Project to review:</strong> {props.courseData.data.find(data => data.id === cr.toReview).projectName} <br />
          <strong>GitHub:</strong>{' '}
          <a href={props.courseData.data.find(data => data.id === cr.toReview).github} target="_blank" rel="noopener noreferrer">
            {props.courseData.data.find(data => data.id === cr.toReview).github}
          </a>
        </p>
      )
    }
    return (
      <p>
        <span>This student is assigned a repo that doesn&#39;t belong to this course to review</span>
        <br />
        <strong>GitHub:</strong>{' '}
        <a href={cr.repoToReview} target="_blank" rel="noopener noreferrer">
          {cr.repoToReview}
        </a>
      </p>
    )
  }

  const renderCodeReview = (i, cr, studentInstance) => {
    const { openWeeks } = state

    return (
      <Accordion fluid styled id={`review${i - 1}`}>
        {' '}
        <Accordion.Title active={openWeeks[i]} index={i} onClick={handleClick}>
          <Icon name="dropdown" /> Code Review {cr.reviewNumber} {cr.points !== null ? ', points ' + cr.points : ''}
        </Accordion.Title>
        <Accordion.Content active={openWeeks[i]}>
          {renderAssignedCodeReviewLink(cr)}
          <strong>Code review:</strong>{' '}
          {cr.linkToReview ? (
            <a href={cr.linkToReview} target="_blank" rel="noopener noreferrer">
              {cr.linkToReview}
            </a>
          ) : (
            'No review linked yet'
          )}
          {cr.points !== null ? <h4>{cr.points} points</h4> : <h4>Not graded yet</h4>}
          <Form onSubmit={gradeCodeReview(cr.reviewNumber, studentInstance)}>
            <label>Points </label>
            <Input name="points" defaultValue={cr.points ? cr.points : ''} type="number" step="0.01" style={{ width: '100px' }} />
            <Input type="submit" value="Grade" />
          </Form>
        </Accordion.Content>
      </Accordion>
    )
  }

  if (state.initialLoading) {
    return <Loader active />
  }

  const createHeaders = (studhead, studentInstance) => {
    let headers = []
    const weekMatcher = i => week => week.weekNumber === i + 1
    // studentInstance is id of student. Type: String
    // Tämä pitää myös korjata.
    const student = studhead.data.find(student => student.id === Number(studentInstance))
    if (student) {
      headers.push(renderStudentCard(student))
      headers.push(
        <span>
          {hasAllReviewsOpen(student) ? (
            <Button type="button" onClick={handleClickHideAll} size="small">
              Hide all reviews
            </Button>
          ) : (
            <Button type="button" onClick={handleClickShowAll(student)} size="small">
              Show all reviews
            </Button>
          )}
          <br />
        </span>
      )
      let i = 0
      let ii = 0
      for (; i < props.selectedInstance.weekAmount; i++) {
        const weeks = student.weeks.find(weekMatcher(i))
        headers.push(renderWeek(i, weeks, studentInstance, false))
      }
      student.codeReviews
        .sort((a, b) => {
          return a.reviewNumber - b.reviewNumber
        })
        .forEach(cr => {
          headers.push(renderCodeReview(i + ii + 1, cr, studentInstance))
          ii++
        })
      if (props.selectedInstance.finalReview) {
        const finalWeek = student.weeks.find(week => week.weekNumber === props.selectedInstance.weekAmount + 1)
        headers.push(renderWeek(i + ii + 1, finalWeek, studentInstance, true))
      }
    }
    return headers.map((header, index) => React.cloneElement(header, { key: index }))
  }

  return (
    <div className="BrowseReviews" style={{ overflowX: 'auto' }}>
      <Loader active={props.loading.loading} />
      {isTeacher() ? (
        <div>
          <BackButton preset="coursePage" />
          <Link to={`/labtool/courses/${props.selectedInstance.ohid}`} style={{ textAlign: 'center' }}>
            <h2> {formatCourseName(props.selectedInstance.name, props.selectedInstance.ohid, props.selectedInstance.start)}</h2>
          </Link>
          {createHeaders(props.courseData, props.studentInstance)}
        </div>
      ) : (
        <p />
      )}
    </div>
  )
}

const mapStateToProps = (state, ownProps) => {
  return {
    ...ownProps,
    user: state.user,
    selectedInstance: state.selectedInstance,
    courseData: state.coursePage,
    studentInstanceToBeReviewed: state.studentInstance,
    loading: state.loading
  }
}

const mapDispatchToProps = {
  createOneComment,
  getOneCI,
  coursePageInformation,
  gradeCodeReview,
  sendEmail,
  resetLoading,
  getCoursesByStudentId,
  updateStudentProjectInfo
}

BrowseReviews.propTypes = {
  courseId: PropTypes.string.isRequired,
  studentInstance: PropTypes.string.isRequired,
  initialLoading: PropTypes.bool,
  location: PropTypes.object,

  user: PropTypes.object.isRequired,
  selectedInstance: PropTypes.object.isRequired,
  courseData: PropTypes.object.isRequired,
  studentInstanceToBeReviewed: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  loading: PropTypes.object.isRequired,

  createOneComment: PropTypes.func.isRequired,
  getOneCI: PropTypes.func.isRequired,
  coursePageInformation: PropTypes.func.isRequired,
  gradeCodeReview: PropTypes.func.isRequired,
  sendEmail: PropTypes.func.isRequired,
  resetLoading: PropTypes.func.isRequired,
  getCoursesByStudentId: PropTypes.func.isRequired,
  updateStudentProjectInfo: PropTypes.func.isRequired
}

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(BrowseReviews)
)
