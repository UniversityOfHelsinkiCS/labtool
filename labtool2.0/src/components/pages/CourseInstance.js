import React from 'react'

const CourseInstance = ({ instance, handleFieldChange }) => {
  return(
    <li> {instance.name} <button onClick={handleFieldChange} name="courseInstanceId" value={instance.id}>Register</button>
    </li>
  )
}

export default CourseInstance