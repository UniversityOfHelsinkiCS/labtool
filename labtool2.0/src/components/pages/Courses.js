import React, { Component } from 'react'
import { Button, List, Container, Header, Table, Label } from 'semantic-ui-react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'

import { getAllCI } from '../../services/courseInstance'

/**
 *  Show all the courses in a single list.
 */
class Courses extends Component {
  componentWillMount() {
    this.props.getAllCI()
  }

  render() {
    return (
      <div>
        <Container>
          <Header as="h2">Courses</Header>
          <Table singleLine color="yellow">
            <Table.Header>
              <Table.Row>
                <Table.HeaderCell colSpan="1"> </Table.HeaderCell>
                <Table.HeaderCell colSpan="1">Course id</Table.HeaderCell>
                <Table.HeaderCell colSpan="1">Course name</Table.HeaderCell>
                <Table.HeaderCell colSpan="1">Course start date</Table.HeaderCell>
                <Table.HeaderCell colSpan="2"> </Table.HeaderCell>
              </Table.Row>
            </Table.Header>

            <Table.Body>
              {this.props.courseInstance.map(instance => (
                <Table.Row key={instance.id}>
                  <Table.Cell>
                    <div>
                      {instance.active === true ? (
                        <Label ribbon style={{ backgroundColor: '#21ba45' }}>
                          Active
                        </Label>
                      ) : (
                          ''
                        )}
                    </div>
                  </Table.Cell>
                  <Table.Cell>{instance.shorterId} </Table.Cell>
                  <Table.Cell>
                    <strong>
                      <a href={`/labtool/courses/${instance.ohid}`}>{instance.name}</a>
                    </strong>
                  </Table.Cell>

                  <Table.Cell> {instance.europeanStart} </Table.Cell>
                  <Table.Cell textAlign="center">
                    <Button circular size="tiny" icon={{ name: 'eye', size: 'large', color: 'blue' }} as={Link} to={`/labtool/courses/${instance.ohid}`} />
                  </Table.Cell>
                </Table.Row>
              ))}
            </Table.Body>
          </Table>

          <div className="Instructions">
            <List>
              <List.Item icon={{ name: 'eye', color: 'blue' }} content="Show course page" />
              <List.Item icon={{ name: 'square', color: 'green' }} content="Course is activated" />
            </List>
          </div>
        </Container>
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    courseInstance: state.courseInstance
  }
}

export default connect(mapStateToProps, { getAllCI })(Courses)
