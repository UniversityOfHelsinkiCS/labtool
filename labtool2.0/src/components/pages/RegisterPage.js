import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import { Form, Input, Grid, Loader } from 'semantic-ui-react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { createStudentCourses, updateStudentProjectInfo } from '../../services/studentinstances'
import { resetLoading, addRedirectHook } from '../../reducers/loadingReducer'
import { Redirect } from 'react-router'
import { getOneCI, coursePageInformation } from '../../services/courseInstance'
import useDebounce from '../../hooks/useDebounce'
import useGithubRepo from '../../hooks/useGithubRepo'
import { GitHubRepoWarning } from './RegisterPage/GitHubRepoWarning'
import DocumentTitle from '../DocumentTitle'
import Error from '../Error'

/**
 * The page user uses to register to a course AS A STUDENT
 */

export const RegisterPage = props => {
  const [repo, setRepo] = useState()
  const [foundExisting, setFoundExisting] = useState(false)
  const [projectName, setProjectName] = useState('')
  const [projectLink, setProjectLink] = useState('https://github.com')
  const debouncedRepo = useDebounce(repo, 500)
  const { githubRepo, error: githubRepoError } = useGithubRepo(debouncedRepo)

  useEffect(() => {
    // run on component mount
    props.resetLoading()
    props.getOneCI(props.courseId)
    props.coursePageInformation(props.courseId)
  }, [])

  useEffect(() => {
    const existing = getExistingData()
    if (!foundExisting && existing.projectLink && !repo) {
      setFoundExisting(true)
      setProjectName(existing.projectName)
      updateRepo(existing.projectLink)
    }
  }, [props.coursePage])

  const getExistingData = () => {
    let projectName = null
    let projectLink = null

    if (props.coursePage && props.coursePage.data) {
      projectName = props.coursePage.data.projectName
      projectLink = props.coursePage.data.github
    }

    return { projectName, projectLink }
  }

  const updateRepo = value => {
    setProjectLink(value)
    const repoLink = value.replace(/^https?:\/\//, '')
    if (repoLink.startsWith('github.com/')) {
      setRepo(repoLink.substring(11).replace(/\/$/, ''))
    } else {
      setRepo(null)
    }
  }

  const handleSubmit = async e => {
    try {
      e.preventDefault()
      const data = {
        // Trim leading and trailing whitespace for accessibility (copy & paste content etc).
        projectName: projectName.trim(),
        github: projectLink.trim(),
        ohid: props.selectedInstance.ohid,
        repoExists: repo !== null ? !githubRepoError : null
      }
      if (props.coursePage && props.coursePage.data !== null) {
        props.addRedirectHook({
          hook: 'STUDENT_PROJECT_INFO_UPDATE_'
        })
        await props.updateStudentProjectInfo(data)
      } else {
        props.addRedirectHook({
          hook: 'STUDENT_COURSE_CREATE_ONE_'
        })
        await props.createStudentCourses(data, props.selectedInstance.ohid)
      }
    } catch (error) {
      console.error(error)
    }
  }

  if (props.loading.redirect) {
    return <Redirect to={`/labtool/courses/${props.selectedInstance.ohid}`} />
  }

  if (props.errors && props.errors.length > 0) {
    return <Error errors={props.errors.map(error => `${error.response.data} (${error.response.status} ${error.response.statusText})`)} />
  }

  return (
    <>
      <DocumentTitle title="Register" />
      <div
        className="RegisterPage"
        style={{
          textAlignVertical: 'center',
          textAlign: 'center'
        }}
      >
        <Loader active={props.loading.loading} inline="centered" />
        <Grid>
          <Grid.Row centered>
            {props.coursePage && props.coursePage.data !== null ? (
              <div>
                <h3>Update your info for {props.selectedInstance.name}</h3>
              </div>
            ) : (
              <div>
                <h3>Register for {props.selectedInstance.name}</h3>
              </div>
            )}
          </Grid.Row>
        </Grid>

        <Grid>
          <Grid.Row centered>
            <Form onSubmit={handleSubmit} className="Register">
              <Form.Group inline>
                <label style={{ width: '100px', textAlign: 'left' }}>Project name</label>
                <Input
                  icon="rocket"
                  iconPosition="left"
                  type="text"
                  className="form-control1"
                  name="projectName"
                  placeholder="MyProjectName"
                  value={projectName}
                  onChange={(e, { value }) => setProjectName(value)}
                  required
                  style={{ minWidth: '30em' }}
                />
              </Form.Group>

              <Form.Group inline>
                <label style={{ width: '100px', textAlign: 'left' }}>GitHub link</label>
                <Input
                  icon="github"
                  iconPosition="left"
                  type="url"
                  className="form-control2"
                  name="github"
                  value={projectLink}
                  onChange={(e, { value }) => updateRepo(value)}
                  required
                  style={{ minWidth: '30em' }}
                />
              </Form.Group>

              {githubRepoError && <GitHubRepoWarning githubRepo={githubRepo} />}

              <Form.Field>
                <button className="ui left floated blue button" type="submit">
                  {' '}
                  Submit
                </button>
                <Link to={`/labtool/courses/${props.selectedInstance.ohid}`}>
                  {' '}
                  <button className="ui right floated button" type="Cancel">
                    Cancel
                  </button>
                </Link>
              </Form.Field>
            </Form>
          </Grid.Row>
        </Grid>
      </div>
    </>
  )
}

const mapStateToProps = (state, ownProps) => {
  return {
    coursePage: state.coursePage,
    selectedInstance: state.selectedInstance,
    courseId: ownProps.courseId,
    loading: state.loading,
    errors: Object.values(state.loading.errors)
  }
}

const mapDispatchToProps = {
  createStudentCourses,
  updateStudentProjectInfo,
  getOneCI,
  coursePageInformation,
  resetLoading,
  addRedirectHook
}

RegisterPage.propTypes = {
  courseId: PropTypes.string.isRequired,

  coursePage: PropTypes.object.isRequired,
  selectedInstance: PropTypes.object.isRequired,
  loading: PropTypes.object.isRequired,

  createStudentCourses: PropTypes.func.isRequired,
  updateStudentProjectInfo: PropTypes.func.isRequired,
  getOneCI: PropTypes.func.isRequired,
  coursePageInformation: PropTypes.func.isRequired,
  resetLoading: PropTypes.func.isRequired,
  addRedirectHook: PropTypes.func.isRequired,

  errors: PropTypes.array
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(RegisterPage)
