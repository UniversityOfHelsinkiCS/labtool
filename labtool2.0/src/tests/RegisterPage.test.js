import React from 'react'
import { RegisterPage } from '../components/pages/RegisterPage'
import { shallow } from 'enzyme'
import { Input } from 'semantic-ui-react'

jest.mock('../hooks/useDebounce', () => {
  return jest.fn().mockImplementation(value => {
    return value
  })
})

jest.mock('../hooks/useGithubRepo', () => {
  return jest.fn().mockImplementation(repo => {
    if (!repo) {
      return { githubRepo: null, error: null }
    }

    return { githubRepo: null, error: 'fake error' }
  })
})

describe('<Register />', () => {
  let wrapper

  const props = {
    selectedInstance: {
      id: 10011,
      name: 'Aineopintojen harjoitustyö: Tietorakenteet ja algoritmit',
      start: '2018-03-11T21:00:00.000Z',
      end: '2018-04-29T21:00:00.000Z',
      active: true,
      weekAmount: 7,
      weekMaxPoints: 3,
      currentWeek: 1,
      ohid: 'TKT20010.2018.K.A.1',
      createdAt: '2018-03-26T00:00:00.000Z',
      updatedAt: '2018-03-26T00:00:00.000Z',
      teacherInstances: [
        {
          id: 10001,
          instructor: false,
          createdAt: '2018-03-26T00:00:00.000Z',
          updatedAt: '2018-03-26T00:00:00.000Z',
          userId: 10010,
          courseInstanceId: 10011,
          firsts: 'Pää',
          lastname: 'Opettaja'
        },
        {
          id: 10011,
          instructor: true,
          createdAt: '2018-03-26T00:00:00.000Z',
          updatedAt: '2018-03-26T00:00:00.000Z',
          userId: 10015,
          courseInstanceId: 10011,
          firsts: 'Ossi Ohjaaja',
          lastname: 'Mutikainen'
        }
      ]
    },
    loading: {
      loading: false,
      loadingHooks: [],
      redirect: false,
      redirectHooks: [],
      redirectFailure: false
    }
  }

  let mockFn = jest.fn()

  beforeEach(() => {
    wrapper = shallow(
      <RegisterPage
        getOneCI={mockFn}
        selectedInstance={props}
        loading={props.loading}
        resetLoading={mockFn}
        courseId={'10012'}
        coursePage={{ data: null }}
        createStudentCourses={mockFn}
        updateStudentProjectInfo={mockFn}
        addRedirectHook={mockFn}
        coursePageInformation={mockFn}
      />
    )
  })

  describe('RegisterPage Component', () => {
    it('is ok', () => {
      true
    })

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render without throwing an error', () => {
      expect(wrapper.find('.Register').length).toEqual(1)
    })

    it('renders project name input', () => {
      expect(wrapper.find('.form-control1').length).toEqual(1)
    })

    it('renders a GitHub link input', () => {
      expect(wrapper.find('.form-control2').length).toEqual(1)
    })

    it('renders a warning if github repo does not exist', () => {
      wrapper
        .find(Input)
        .find({ name: 'github' })
        .simulate('change', null, { value: 'https://github.com/invalid_repo' })

      expect(wrapper.find('GitHubRepoWarning').length).toEqual(1)
    })
  })
})
