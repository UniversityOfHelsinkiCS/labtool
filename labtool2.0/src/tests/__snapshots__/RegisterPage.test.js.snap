// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Register /> RegisterPage Component should render correctly 1`] = `
<div
  className="RegisterPage"
  style={
    Object {
      "textAlign": "center",
      "textAlignVertical": "center",
    }
  }
>
  <Loader
    active={false}
    inline="centered"
  />
  <Grid>
    <GridRow
      centered={true}
    >
      <div>
        <h3>
          Register for 
        </h3>
      </div>
    </GridRow>
  </Grid>
  <Grid>
    <GridRow
      centered={true}
    >
      <Form
        as="form"
        onSubmit={[Function]}
      >
        <FormGroup
          inline={true}
        >
          <label
            style={
              Object {
                "textAlign": "left",
                "width": "100px",
              }
            }
          >
            Project name
          </label>
          <Input
            className="form-control1"
            icon="rocket"
            iconPosition="left"
            name="projectName"
            placeholder="MyProjectName"
            required={true}
            style={
              Object {
                "minWidth": "30em",
              }
            }
            type="text"
          />
        </FormGroup>
        <FormGroup
          inline={true}
        >
          <label
            style={
              Object {
                "textAlign": "left",
                "width": "100px",
              }
            }
          >
            GitHub link
          </label>
          <Input
            className="form-control2"
            icon="github"
            iconPosition="left"
            name="github"
            placeholder="https://github.com/myaccount/myrepo"
            required={true}
            style={
              Object {
                "minWidth": "30em",
              }
            }
            type="url"
          />
        </FormGroup>
        <FormField>
          <button
            className="ui left floated blue button"
            type="submit"
          >
             
            Submit
          </button>
          <Link
            replace={false}
            to="/labtool/courses/undefined"
          >
             
            <button
              className="ui right floated button"
              type="Cancel"
            >
              Cancel
            </button>
          </Link>
        </FormField>
      </Form>
    </GridRow>
  </Grid>
</div>
`;
